### ISSUE TITLE
lazy-require instantiates module twice

### ISSUE THREAD

**Issue by m4burns** [racket/racket#3805]
**What version of Racket are you using?**
8.0 [cs], 8.1.0.2 [cs] snapshot 20210427-6e3a94b6ac

**What program did you run?**
main.rkt:
```
#lang racket

(module bar racket
  (require racket/lazy-require)
  (provide bar)
  (lazy-require
    ["foo.rkt" (foo)])
  (define bar foo))

(module baz racket
  (require racket/lazy-require)
  (provide baz)
  (lazy-require
    ["foo.rkt" (foo)])
  (define baz foo))

(require (submod "." bar)
         (submod "." baz))

(for-each thread-wait
  (map thread (list baz bar)))
```

foo.rkt:
```
#lang racket

(provide foo)

(define (foo)
  (printf "hello~n"))
```

**What should have happened?**
```
$ racket main.rkt
hello
hello
```

**If you got an error message, please include it here.**
```
$ racket main.rkt
hello
define-values: assignment disallowed;
 cannot re-define a constant
  constant: foo
  in module:"/home/marc/foo.rkt"
  context...:
   body of "/home/marc/foo.rkt"
   /usr/local/racket/collects/racket/lazy-require.rkt:89:13: get-sym
   /usr/local/racket/collects/racket/promise.rkt:65:10
   /usr/local/racket/collects/racket/promise.rkt:45:2
   /usr/local/racket/collects/racket/lazy-require.rkt:106:6
```

**Please include any other relevant details**
Linux x86_64

It seems that calling two functions bound by `lazy-require` in a race on two threads can cause a module to be instantiated twice. I found this when trying to establish database connections in two distinct threads at roughly the same time - the connection constructors for each database implementation are `lazy-require`d.

**Comment by m4burns** [comment]
When I encountered this, the error message was different:

```
instantiate-linklet: mismatch;
 reference to a variable that is not exported;
 possibly, bytecode file needs re-compile because dependencies changed
  name: maybe-add-state-undo
  exporting instance: "/usr/local/racket/collects/syntax/parse/private/residual.rkt"
  importing instance: "/usr/local/racket/collects/syntax/parse/private/lib.rkt"
  context...:
   body of "/usr/local/racket/collects/syntax/parse/private/lib.rkt"
   .../racket/lazy-require.rkt:89:13: get-sym
   /usr/local/racket/collects/racket/promise.rkt:65:10
   /usr/local/racket/collects/racket/promise.rkt:45:2
   /usr/local/racket/collects/racket/lazy-require.rkt:106:6
   /usr/local/racket/collects/racket/contract/private/arrow-val-first.rkt:555:3
   /home/marc/projects/kn0x/hot-jockey/transaction-stash.rkt:14:0: with-connection
   /home/marc/projects/kn0x/hot-jockey/main.rkt:122:0: main
   /usr/local/racket/collects/racket/logging.rkt:43:0: with-intercepted-logging
   /home/marc/projects/kn0x/racket-common/service.rkt:47:8
```

I removed all `compiled` directories outside the racket installation and got the same message. I then managed to cause the error shown in the issue. I'm guessing these two errors share the same underlying cause.

**Comment by m4burns** [comment]
This is probably the same issue as #3128 . I can't easily remove the threads in the application, which looks something like:

```
(define service1
  (thread
    (thunk
      ....
      (sqlite3-connect ...)
      ....)))

(define service2
  (thread
    (thunk
      ....
      (postgresql-connect ...)
      ....)))

(sync service1 service2)
```

**Comment by m4burns** [comment]
#3809 causes the program in the original post to run as I expected and resolves the issue in our application.

**Comment by m4burns** [comment]
Confirmed resolved by d9c128fe656516777b7d653ab2d18519f6603e4f
